### DISCUSSION TITLE
Add loaded client side page to provisioning template (Types mismatch)

### DISCUSSION THREAD

**Discussion by DmitriyVdE** [pnp/pnpcore#1172] [Q&A] [answered]
Hi

I'm currently making some provisioning templates and wanted to automate part of the process for future use but I'm a bit stuck on the client side pages part.

I'm currently attempting to load pages using `context.Web.LoadClientSidePageAsync(pageName);` using a `ClientContext`, not a `PnPContext`, and this returns a `IPage` object. This is a `PnP.Core.Model.SharePoint.IPage` but the provisioning template I created using `ProvisioningTemplate template = context.Web.GetProvisioningTemplate(ptci);` expects a `PnP.Framework.Provisioning.Model.ClientSidePage` in its `template.ClientSidePages.Add(PAGE)` function.

The issue is that I can't seems to find a way to either load the page in the correct type or convert between the types.

My current 'fix' is to just create an export using the `Export-PnPPage` command using `PnP.PowerShell` and then manually add the XML to the template file.

Thanks in advance
-Dmitriy

**Comment by jansenbe** [comment]
Hey @DmitriyVdE , these indeed are completely different things. PnP Core SDK contains the pages API which has an IPage representing a page and all actions/settings for that page. The Provisioning Engine in PnP Core SDK allows you to define pages in a provisioning template and for that a `ClientSidePage` is needed. If you want to programmatically add pages to your template that I would suggest looking at the code in PnP PowerShell (https://github.com/pnp/powershell/blob/dev/src/Commands/Pages/ExportPage.cs#L68-L101) and PnP Framework (https://github.com/pnp/pnpframework/blob/dev/src/lib/PnP.Framework/Provisioning/ObjectHandlers/Utilities/ClientSidePageContentsHelper.cs).

**Reply by DmitriyVdE** [reply]
Thanks @jansenbe for your fast reply.
I'm already using PnP PowerShell since it indeed seemed like the quickest route for what I'm currently doing.

And using PnP Framework would probably also require me to use the PnP Framework authentication yes?

**Reply by jansenbe** [reply, marked as answer]
@DmitriyVdE : you can use the `AuthenticationManager` class for auth in PnP Framework, under the covers PnP Core SDK is used by PnP Framework to load the page but we handle auth automatically for you.
